src/masternodes/mn_rpc.cpp

```diff
@@ -3768,6 +3769,213 @@ UniValue sendtokenstoaddress(const JSONRPCRequest& request) {
+
+    const uint256 hash(ParseHashV(request.params[0], "txid"));
+
+    CTransactionRef tx;
+    uint256 hashBlock;
+
+    // Search wallet if available
+    if (pwallet) {
+        LOCK(pwallet->cs_wallet);
+        auto it = pwallet->mapWallet.find(hash);
+        if (it != pwallet->mapWallet.end())
+        {
+            tx = it->second.tx;
+            hashBlock = it->second.hashBlock;
+        }
+    }
+
+    CBlockIndex* blockindex{nullptr};
+
+    // No wallet or not a wallet TX, try mempool, txindex and a block if hash provided
+    if (!pwallet || !tx)
+    {
+        if (!request.params[1].isNull()) {
+            LOCK(cs_main);
+
+            uint256 blockhash = ParseHashV(request.params[1], "blockhash");
+            blockindex = LookupBlockIndex(blockhash);
+            if (!blockindex) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block hash not found");
+            }
+        }
+
+        bool f_txindex_ready{false};
+        if (g_txindex && !blockindex) {
+            f_txindex_ready = g_txindex->BlockUntilSyncedToCurrentChain();
+        }
+
+        if (!GetTransaction(hash, tx, Params().GetConsensus(), hashBlock, blockindex)) {
+            std::string errmsg;
+            if (blockindex) {
+                if (!(blockindex->nStatus & BLOCK_HAVE_DATA)) {
+                    throw JSONRPCError(RPC_MISC_ERROR, "Block not available");
+                }
+                errmsg = "No such transaction found in the provided block.";
+            } else if (!g_txindex) {
+                errmsg = "No such mempool or wallet transaction. Use -txindex or provide a block hash.";
+            } else if (!f_txindex_ready) {
+                errmsg = "No such mempool or wallet transaction. Transactions are still in the process of being indexed.";
+            } else {
+                errmsg = "No such mempool, wallet or blockchain transaction.";
+            }
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, errmsg);
+        }
+    }
+
+    int nHeight{0};
+    bool actualHeight{false};
+    CustomTxType guess;
+    UniValue txResults(UniValue::VOBJ);
+    Res res{};
+
+    if (tx)
+    {
+        LOCK(cs_main);
+
+        // Found a block hash but no block index yet
+        if (!hashBlock.IsNull() && !blockindex) {
+            blockindex = LookupBlockIndex(hashBlock);
+        }
+
+        // Default to next block height
+        nHeight = ::ChainActive().Height() + 1;
+
+        // Get actual height if blockindex avaiable
+        if (blockindex) {
+            nHeight = blockindex->nHeight;
+            actualHeight = true;
+        }
+
+        // Skip coinbase TXs except for genesis block
+        if ((tx->IsCoinBase() && nHeight > 0)) {
+            return "Coinbase transaction. Not a custom transaction.";
+        }
+
+        std::vector<unsigned char> metadata;
+        guess = GuessCustomTxType(*tx, metadata);
+        CCustomCSView mnview_dummy(*pcustomcsview);
+
+        switch (guess)
```

Comment:
Can we move this switch section to a seperate function, this function is quite long.